docs/reqcontext.rst

```diff
@@ -175,6 +200,45 @@ Certain parts of the test system might also temporarily create a request
 context without calling the before-request handlers.  Make sure to write
 your teardown-request handlers in a way that they will never fail.
 
+.. _ensuring-teardown-callbacks-succeed:
+
+Ensuring teardown callbacks succeed
+-----------------------------------
+
+.. admonition:: Watch Out
+
+    Teardown callbacks must never raise. If they do, teardown callbacks that
+    haven't run yet won't run, and request/app contexts won't pop, which will
+    cause memory/file leaks at best and context persisting into later requests
+    at worst.
+
+    Note that extensions can register teardown callbacks, and those callbacks
```

Comment:
I agree that section feels a bit off-putting to someone new to Flask, and yet rings true enough that probably some code changes are in order to help ameliorate this issue.

That said, I'm personally +1 on merging as is, as long as we also create a new issue to track potential code improvements. I'd hate to see anyone else get bit by this and waste hours of their life until we get around to deciding what to do on the code side.

@ThiefMaster your thoughts? Is there a better short-term solution?